### PR TITLE
Webauthn jwk formats

### PR DESCRIPTION
**Pull request by edmundnoble**
Still to do:
- [x] fork tests of webauthn keys + sigs
- [x] fork new sig format
- [ ] fix rosetta or determine if we need to
- [ ] partial replay at the minimum from when webauthn signatures became valid on mainnet

### CODE REVIEW

File: bench/Chainweb/Pact/Backend/ForkingBench.hs [hunk 1]
```diff
@@ -263,7 +263,7 @@ data Resources
     , blockHeaderDb :: !BlockHeaderDb
     , pactService :: !(Async (), PactQueue)
     , mainTrunkBlocks :: ![T3 ParentHeader BlockHeader PayloadWithOutputs]
-    , coinAccounts :: !(MVar (Map Account (NonEmpty Ed25519KeyPairCaps)))
+    , coinAccounts :: !(MVar (Map Account (NonEmpty (DynKeyPair, [SigCapability]))))
```
Comment (jwiegley) [inline bench/Chainweb/Pact/Backend/ForkingBench.hs:266]: To avoid repetition, could we change this tuple into a proper sum type?

File: bench/Chainweb/Pact/Backend/ForkingBench.hs [hunk 2]
```diff
@@ -484,12 +485,16 @@ createCoinAccounts v meta = traverse (go <*> createCoinAccount v meta) names
...
+formatB16PubKey :: DynKeyPair -> Text
+formatB16PubKey = \case
+  DynEd25519KeyPair kp -> toB16Text $ getPublic kp
+  DynWebAuthnKeyPair pub _ _ -> toB16Text $ exportWebAuthnPublicKey pub
 
 safeCapitalize :: String -> String
 safeCapitalize = maybe [] (uncurry (:) . bimap toUpper (Prelude.map toLower)) . Data.List.uncons
 
+
+-- TODO: Use the new `assertCommand` function.
```
Comment (jwiegley) [inline bench/Chainweb/Pact/Backend/ForkingBench.hs:497]: Will this TODO be addressed here? If not, let's make a ticket.

File: src/Chainweb/Pact/Utils.hs [hunk 3]
```diff
@@ -64,15 +63,16 @@ toTxCreationTime (Time timespan) =
   TxCreationTime $ ParsedInteger $ fromIntegral $ timeSpanToSeconds timespan
 
 
-validatePubKey :: P.PublicKeyText -> Bool
-validatePubKey = validateKeyFormat
 
+-- TODO: This will fail for k: accounts that correspond to WebAuthn public
+-- keys. Consider extending it when we integrate WebAuthn with k: accounts.
```
Comment (jwiegley) [inline src/Chainweb/Pact/Utils.hs:68]: Also here: TODO comments are usually just a record of a thought, rarely acted upon. If you really want to make this happen, Asana is the place!

File: src/Chainweb/Transaction.hs [hunk 4]
```diff
@@ -32,7 +32,7 @@ import Control.Lens
 import qualified Data.Aeson as Aeson
 import Data.ByteString.Char8 (ByteString)
 import qualified Data.ByteString.Char8 as B
-import qualified Data.ByteString.Short as SB
+import qualified Data.ByteString.Short as SBS
```
Comment (jwiegley) [inline src/Chainweb/Transaction.hs:35]: This change is completely unrelated to this PR, right?

File: src/Chainweb/Pact/RestAPI/Server.hs [hunk 5]
```diff
@@ -679,13 +686,26 @@ internalPoll pdb bhdb mempool pactEx cut confDepth requestKeys0 = do
...
+    let pwt = mkPayloadWithText cmdBs parsedPact
+        commandParsed = cmdText { _cmdPayload = pwt }
+    in
+        if assertCommand
+             commandParsed
+             (validPPKSchemes v cid maxBound)
+             (isWebAuthnPrefixLegal v cid maxBound)
+        then Right commandParsed
+        else Left "Command failed validation"
+  Left e -> Left $ "Pact parsing error: " ++ e
```
Comment (emilypi) [inline src/Chainweb/Pact/RestAPI/Server.hs:702]: 👍

File: src/Chainweb/Rosetta/Utils.hs [hunk 6]
```diff
@@ -762,6 +762,8 @@ getCmdPayload (Command p _ _) =
     (decodeStrict' $! T.encodeUtf8 p)
 
 
+-- TODO: This assumes Rosettas signatures are all Ed25519 signatures
+-- (Not webauthn).
```
Comment (emilypi) [inline src/Chainweb/Rosetta/Utils.hs:766]: Per Linda: we're not looking to continue to support Rosetta, so please feel free to do what you need to in order to make rosetta work here.

File: tools/ea/Ea.hs [hunk 7]
```diff
@@ -182,6 +182,11 @@ mkChainwebTxs' :: [Command Text] -> IO [ChainwebTransaction]
 mkChainwebTxs' rawTxs =
     forM rawTxs $ \cmd -> do
         let cmdBS = fmap TE.encodeUtf8 cmd
+            -- TODO: Use the new `assertCommand` function.
+            -- We want to delete `verifyCommand` at some point.
+            -- It's not critical for Ea because WebAuthn signatures (which
+            -- the new `assertCommand` knows how to handle) are not present
+            -- in Genesis blocks.
```
Comment (emilypi) [inline tools/ea/Ea.hs:189]: 👍

File: src/Chainweb/Pact/TransactionExec.hs [hunk 8]
```diff
@@ -784,6 +786,9 @@ enablePact48 v cid bh = [FlagDisablePact48 | not (chainweb220Pact v cid bh)]
 enablePact49 :: ChainwebVersion -> V.ChainId -> BlockHeight -> [ExecutionFlag]
 enablePact49 v cid bh = [FlagDisablePact49 | not (chainweb221Pact v cid bh)]
 
+enablePact410 :: ChainwebVersion -> V.ChainId -> BlockHeight -> [ExecutionFlag]
+enablePact410 v cid bh = [FlagDisablePact410 | not (chainweb222Pact v cid bh)]
+
 -- | Even though this is not forking, abstracting for future shutoffs
 disableReturnRTC :: ChainwebVersion -> V.ChainId -> BlockHeight -> [ExecutionFlag]
 disableReturnRTC _v _cid _bh = [FlagDisableRuntimeReturnTypeChecking]
```
Comment (emilypi) [inline src/Chainweb/Pact/TransactionExec.hs:794]: this needs to similarly hook up the webauthn signature validation for `applyLocal`, no?

File: bench/Chainweb/Pact/Backend/ForkingBench.hs (hunk 2 above)
Comment (imalsogreg) [reply bench/Chainweb/Pact/Backend/ForkingBench.hs:497]: https://app.asana.com/0/1204132908184958/1206014030451021/f

File: src/Chainweb/Pact/Utils.hs (hunk 3 above)
Comment (imalsogreg) [reply src/Chainweb/Pact/Utils.hs:68]: https://app.asana.com/0/1204132908184958/1206014030451021/f

File: bench/Chainweb/Pact/Backend/ForkingBench.hs [hunk 9]
```diff
@@ -450,6 +450,8 @@ formatB16PubKey = toB16Text . getPublic
...
```
[the remark this answers is not in the document]
Comment (imalsogreg) [reply bench/Chainweb/Pact/Backend/ForkingBench.hs:497]: Made an Asana ticket

File: src/Chainweb/Transaction.hs (hunk 4 above)
Comment (imalsogreg) [reply src/Chainweb/Transaction.hs]: Thx, fixed!

File: bench/Chainweb/Pact/Backend/ForkingBench.hs (hunk 1 above)
Comment (imalsogreg) [reply bench/Chainweb/Pact/Backend/ForkingBench.hs:266]: Deferring to later, adding to the webauthn cleanup asana ticket.

**Review by emilypi — APPROVED** [review]

File: src/Chainweb/Rosetta/RestAPI/Server.hs [hunk 10]
```diff
@@ -345,17 +345,18 @@ constructionParseH
...
     -> ConstructionParseReq
     -> Handler ConstructionParseResp
-constructionParseH v (ConstructionParseReq net isSigned tx) =
+constructionParseH v (ConstructionParseReq net isSigned tx) = do
   either throwRosettaError pure work
   where
     work :: Either RosettaError ConstructionParseResp
     work = do
+      cid <- annotate rosettaError' (validateNetwork v net)
       void $ annotate rosettaError' (validateNetwork v net)
```
Comment (edmundnoble) [inline src/Chainweb/Rosetta/RestAPI/Server.hs:354]: You don't need the `void $ ...` anymore.

File: test/pact/contTXOUTNew.pact [hunk 11]
```diff
@@ -25,9 +25,9 @@
       }
     , "events": []
     , "gas": 6
-    , "logs": "9JwtbUnJnZVC62lihy1YDA7Gmi2Mkopl62GkMGB3QZI"
+    , "logs": "c_Xbqb7BZetrHBtU7nGdZ4yoc1_Y1iE3CgsO03gOrbc"
     , "meta": {}
-    , "req-key": "rwFh_qoxumclpxZj6QIo5GOM0CmE5AztlKukhFCRTVA"
+    , "req-key": "pHsX8RJkJNRlrGflD_vRiKOsjlLSVGdf1hYU_eI6fgU"
```
Comment (jmcardon) [inline test/pact/contTXOUTNew.pact:30]: Note: these req keys and logs changed because we added `NetworkId` to the command.

File: test/golden/transfer-accounts-expected.txt [hunk 12]
```diff
@@ -3,4 +3,4 @@
 - cmd:
     contents: (test1.transfer "Acct1" "Acct2" 1.00)
     tag: Code
-  output: noxWe0aOpl3J20QHMVJP8EvPGG8kftv63GeIjvJwrXQ
+  output: O4tU1VdTniyZH8CkVPKViCvHT6C_lke6lmObZeKjPL8
```
Comment (jmcardon) [inline test/golden/transfer-accounts-expected.txt:6]: Ditto here,  no red flags, just networkId difference.

**Review by jmcardon — APPROVED** [review]
